### PR TITLE
rgw/multisite: handle shard_id correctly in data sync when bucket num_shards is 0

### PR DESCRIPTION
**Pull request by smanjara**
For buckets that have num_shards set to 0, bucket instance
will not have a shard_id delimiter. When this bucket instance is parsed,
we end up assigning a value of -1 to shard_id, which is invalid
in data sync. This change ensures that we represent the shard_id correctly
by giving it a valid number

https://tracker.ceph.com/issues/59489

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/services/svc_bi_rados.cc [hunk 1]
```diff
@@ -207,10 +207,10 @@ int RGWSI_BucketIndex_RADOS::open_bucket_index(const DoutPrefixProvider *dpp,
     return ret;
   }
 
-  get_bucket_index_objects(bucket_oid_base, idx_layout.layout.normal.num_shards,
+  get_bucket_index_objects(bucket_oid_base, rgw::num_shards(idx_layout.layout.normal),
                            idx_layout.gen, bucket_objs, shard_id);
   if (bucket_instance_ids) {
-    get_bucket_instance_ids(bucket_info, idx_layout.layout.normal.num_shards,
+    get_bucket_instance_ids(bucket_info, rgw::num_shards(idx_layout.layout.normal),
```
Comment (cbodley) [inline src/rgw/services/svc_bi_rados.cc:213]: ..but won't this break access to existing buckets that had num_shards=0?

**Review by cbodley — COMMENTED** [review]
the change to `init_index()` looks correct..

**Comment by cbodley** [comment]
since we can't change the shard object names of existing buckets in `open_bucket_index()`, i don't think this is really going to fix anything. if num_shards=0 is causing other bugs, we need to root-cause and fix those instead

File: src/rgw/services/svc_bi_rados.cc (hunk 1 above)
Comment (smanjara) [reply src/rgw/services/svc_bi_rados.cc:213]: yeah I am still testing this.

**Comment by smanjara** [comment]
> > since we can't change the shard object names of existing buckets in `open_bucket_index()`, i don't think this is really going to fix anything. if num_shards=0 is causing other bugs, we need to root-cause and fix those instead
> 
> @cbodley for a simple case of a bucket with num_shards 0 (without this fix), an incremental sync causes radosgw to crash:
> 
> ```
>  8: (RGWSyncBucketCR::operate(DoutPrefixProvider const*)+0x1e84) [0x55aa93b1cb92]
>  9: (RGWCoroutine::operate_wrapper(DoutPrefixProvider const*)+0xa) [0x55aa936271de]
>  10: (RGWCoroutinesStack::operate(DoutPrefixProvider const*, RGWCoroutinesEnv*)+0x118) [0x55aa937727a8]
>  11: (RGWCoroutinesManager::run(DoutPrefixProvider const*, std::__cxx11::list<RGWCoroutinesStack*, std::allocator<RGWC
> oroutinesStack*> >&)+0x12e) [0x55aa937731b4]
>  12: (RGWCoroutinesManager::run(DoutPrefixProvider const*, RGWCoroutine*)+0x84) [0x55aa9377400a]
>  13: (RGWRemoteDataLog::run_sync(DoutPrefixProvider const*, int)+0x85) [0x55aa93b216dd]
>  14: (RGWDataSyncProcessorThread::process(DoutPrefixProvider const*)+0x46) [0x55aa93974760]
> ```
> 
> but you are right that the older buckets will still have the same shard object names and an incremental sync can still result in the crash. Looking further into what's causing it.

seems to be crashing when CheckBucketShardStatusIsIncremental() is called. specifically at:
https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_data_sync.cc#L3382

**Comment by smanjara** [comment]
> > > since we can't change the shard object names of existing buckets in `open_bucket_index()`, i don't think this is really going to fix anything. if num_shards=0 is causing other bugs, we need to root-cause and fix those instead
> > 
> > 
> > @cbodley for a simple case of a bucket with num_shards 0 (without this fix), an incremental sync causes radosgw to crash:
> > ```
> >  8: (RGWSyncBucketCR::operate(DoutPrefixProvider const*)+0x1e84) [0x55aa93b1cb92]
> >  9: (RGWCoroutine::operate_wrapper(DoutPrefixProvider const*)+0xa) [0x55aa936271de]
> >  10: (RGWCoroutinesStack::operate(DoutPrefixProvider const*, RGWCoroutinesEnv*)+0x118) [0x55aa937727a8]
> >  11: (RGWCoroutinesManager::run(DoutPrefixProvider const*, std::__cxx11::list<RGWCoroutinesStack*, std::allocator<RGWC
> > oroutinesStack*> >&)+0x12e) [0x55aa937731b4]
> >  12: (RGWCoroutinesManager::run(DoutPrefixProvider const*, RGWCoroutine*)+0x84) [0x55aa9377400a]
> >  13: (RGWRemoteDataLog::run_sync(DoutPrefixProvider const*, int)+0x85) [0x55aa93b216dd]
> >  14: (RGWDataSyncProcessorThread::process(DoutPrefixProvider const*)+0x46) [0x55aa93974760]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but you are right that the older buckets will still have the same shard object names and an incremental sync can still result in the crash. Looking further into what's causing it.
> 
> seems to be crashing when CheckBucketShardStatusIsIncremental() is called. specifically at: https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_data_sync.cc#L3382

huh actually no. it happens right when we enter the incremental sync at https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_data_sync.cc#L5864 
we are asserting if shard_id is non-negative and the shard_id we pass happens to be -1.

**Comment by smanjara** [comment]
so it comes down to the way we parse bucket instance without the shard id delimiter in
https://github.com/ceph/ceph/blob/main/src/rgw/rgw_bucket.cc#L57 where shard_id gets the value of  '-1'.

**Comment by cbodley** [comment]
yeah, i think it makes sense for data sync to treat -1 as 1 shard

**Comment by smanjara** [comment]
@cbodley please review.

**Comment by smanjara** [comment]
https://pulpito.ceph.com/cbodley-2023-04-21_14:36:40-rgw-wip-smanjara-testing-distro-default-smithi/

**Comment by smanjara** [comment]
@cbodley could you review this one please?

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
@smanjara can you please make sure these fixes make it to reef?

**Pull request merged**